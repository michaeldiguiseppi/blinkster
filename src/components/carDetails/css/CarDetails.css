

.no-list-style {
    list-style-type: none;
}

.vehicle-body-style {
    text-transform: capitalize;
}

#card-overflow-scroll {
    overflow: scroll;
}