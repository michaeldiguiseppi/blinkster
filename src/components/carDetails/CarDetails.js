import React, { Component } from 'react';
import moment from 'moment';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import * as carsActions from '../../actions/cars';
import './css/CarDetails.css';

export class CarDetails extends Component {
    constructor(props) {
        super(props);

        this._fetchCar = this._fetchCar.bind(this);
        this._renderCar = this._renderCar.bind(this);
    }

    componentDidMount() {
        this._fetchCar();
    }

    _fetchCar() {
        let { carsActions, match: { params } } = this.props;
        return carsActions.fetchCar(params.year, params.make, params.model);
    }

    _renderCar() {
        let { car } = this.props.cars;
        return (
            <div>   
            { car && Object.keys(car).length !== 0 ? 
            <div>
                <div className="vehicle-img mx-auto d-block">
                    <img src={`${car.image_url}`} className="img-fluid rounded" alt={`${car.year} ${car.make} ${car.model}`} />
                </div>
                <div className="vehicle-info">
                    <div className="card">
                        <h3 className="card-header text-center">{ `${car.year} ${car.make} ${car.model}` }</h3>
                        <div className="card-body">
                            <ul className="no-list-style">
                                <li className="veh-info">Year: { car.year }</li>
                                <li className="veh-info">Make: { car.make }</li>
                                <li className="veh-info">Model: { car.model }</li>
                                <li className="veh-info">Mileage: { car.mileage }</li>
                                <li className="veh-info">Drivetrain: { car.drivetrain ? car.drivetrain : "N/A" }</li>
                                <li className="vehicle-body-style veh-info">Body Style: { car.bodytype }</li>
                                <li className="veh-info">Posted: { `${moment(car.created_at).format("MMMM Do, YYYY")}` }</li>
                            </ul>
                        </div>
                    </div>
                </div>
            </div> 
            : 
            <div className="jumbotron text-center">Car not found</div>}
            </div>
        )
    }

    render() {
        return (
            <div>
                { this._renderCar() }
            </div>
        );
    }
}

function mapStateToProps(state) {
    let cars = state.cars || [];

    return {
        cars: cars,
    }
}

function mapDispatchToProps(dispatch) {
    return {
        carsActions: bindActionCreators(carsActions, dispatch)
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(CarDetails);