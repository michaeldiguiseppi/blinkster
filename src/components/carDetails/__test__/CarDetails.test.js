import React from 'react';
import { CarDetails } from '../CarDetails';
import renderer from 'react-test-renderer';
import { shallow, configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });


describe('<CarDetails /> component', () => {
    let carDetailsComponent;
    const defaultProps = {
        cars: {
            isLoading: false,
            car: {
                "year" : 2013,
                "make" : "Kia",
                "model" : "Optima",
                "mileage" : 24235,
                "drivetrain" : "FWD",
                "bodytype" : "sedan",
                "image_url" : "http://www.optimaforums.com/forum/attachments/new-member-introductions/11137d1347548855-new-2013-kia-optima-sx-l-titanium-photo.jpg",
                "created_at" : "2016-10-14T20:13:22.586Z"
              }
        },
        carsActions: {
            fetchCar: () => {}
        },
        match: {
            params: {
                index: 0
            }
        },
    }

    beforeEach(() => {
        carDetailsComponent = mount(<CarDetails {...defaultProps} />);
    });

    afterEach(() => {
        carDetailsComponent.unmount();
    });

    it('should render without breaking', () => {
        expect(carDetailsComponent.exists()).toBe(true);
    });

    it('should match the snapshot', () => {
        const tree = renderer
            .create(<CarDetails {...defaultProps} />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should have a list of car information', () => {
        expect(carDetailsComponent.find('ul.no-list-style')).toHaveLength(1);
        // Should have seven key pieces of information rendered
        expect(carDetailsComponent.find('.veh-info')).toHaveLength(7);
    });

    it('should render an image of the car', () => {
        expect(carDetailsComponent.find('img')).toHaveLength(1);
        expect(carDetailsComponent.find('img').prop('src')).toEqual(defaultProps.cars.car.image_url);
    });

    it('should render an error message if no car is found', () => {
        let props = {
            cars: {
                isLoading: false,
                car: {},
            },
            carsActions: {
                fetchCar: () => {}
            },
            match: {
                params: {
                    index: 0
                }
            },
        }
        carDetailsComponent.unmount();
        carDetailsComponent = mount(<CarDetails {...props} />);
        let jumbotron = carDetailsComponent.find('div.jumbotron');
        expect(carDetailsComponent.exists()).toBe(true);
        expect(jumbotron).toHaveLength(1);
        expect(jumbotron.text()).toEqual("Car not found");
    });
});