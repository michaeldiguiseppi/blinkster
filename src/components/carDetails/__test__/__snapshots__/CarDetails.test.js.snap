// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CarDetails /> component should match the snapshot 1`] = `
<div>
  <div>
    <div>
      <div
        className="vehicle-info"
      >
        <div
          className="card w-75 mx-auto"
        >
          <h3
            className="card-header text-center"
          >
            2013 Kia Optima
          </h3>
          <img
            alt="2013 Kia Optima"
            className="card-img-top"
            src="http://www.optimaforums.com/forum/attachments/new-member-introductions/11137d1347548855-new-2013-kia-optima-sx-l-titanium-photo.jpg"
          />
          <div
            className="card-body"
            id="card-overflow-scroll"
          >
            <ul
              className="no-list-style card-text"
            >
              <li
                className="veh-info"
              >
                Year: 
                2013
              </li>
              <li
                className="veh-info"
              >
                Make: 
                Kia
              </li>
              <li
                className="veh-info"
              >
                Model: 
                Optima
              </li>
              <li
                className="veh-info"
              >
                Mileage: 
                24235
              </li>
              <li
                className="veh-info"
              >
                Drivetrain: 
                FWD
              </li>
              <li
                className="vehicle-body-style veh-info"
              >
                Body Style: 
                sedan
              </li>
              <li
                className="veh-info"
              >
                Posted: 
                October 14th, 2016
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
